addons/pos_pinelabs/static/src/app/payment_pinelabs.js

```diff
@@ -0,0 +1,245 @@
+import { _t } from "@web/core/l10n/translation";
+import { PaymentInterface } from "@point_of_sale/app/payment/payment_interface";
+import { AlertDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
+import { serializeDateTime } from "@web/core/l10n/dates";
+
+const REQUEST_TIMEOUT = 10000;
+const CANCEL_REQUEST_TIME_LIMIT = 125000;
+const { DateTime } = luxon;
+
+export class PaymentPinelabs extends PaymentInterface {
+    setup() {
+        super.setup(...arguments);
+        this.paymentInitialTime = null;
+        this.pollingTimeout = null;
+        this.inactivityTimeout = null;
+        this.queued = false;
+        this.payment_stopped = false;
+    }
+
+    send_payment_request(cid) {
+        super.send_payment_request(cid);
+        return this._processPinelabs();
+    }
+
+    pendingPinelabsPaymentLine() {
+        return this.pos.getPendingPaymentLine("pinelabs");
+    }
+
+    send_payment_cancel(order, cid) {
+        super.send_payment_cancel(order, cid);
+        return this._pinelabsCancel();
+    }
+
+    _callPinelabs(data, action) {
+        return this.env.services.orm.silent
+            .call("pos.payment.method", action, [[this.payment_method_id.id], data])
+            .catch(this._handle_odoo_connection_failure.bind(this));
+    }
+
+    _handle_odoo_connection_failure(data = {}) {
+        // handle timeout
+        const line = this.pendingPinelabsPaymentLine();
+        if (line) {
+            line.set_payment_status("retry");
+        }
+        this._showError(
+            _t(
+                "Unable to connect to the Odoo server. Please check your internet connection and try again."
+            )
+        );
+
+        return Promise.reject(data); // prevent subsequent unFulFilled's from being called
+    }
+
+    /**
+     * This method processes the response from Pinelabs
+     * when we request payment or cancellation.
+     */
+    _pinelabsHandleResponse(response) {
+        const line = this.pendingPinelabsPaymentLine();
+        if (response.error) {
+            line.set_payment_status("retry");
+            this.payment_stopped
+                ? this._showError(_t("Transaction failed due to inactivity"))
+                : this._showError(response.error);
+            this._removePaymentHandler(["transactionNumber", "plutusTransactionReferenceID"]);
+            return Promise.resolve(false);
+        }
+        line.set_payment_status("waitingCard");
+        localStorage.setItem("plutusTransactionReferenceID", response.plutusTransactionReferenceID);
+        return this._waitForPaymentToConfirm();
+    }
+
+    _pinelabsCancel() {
+        const line = this.pendingPinelabsPaymentLine();
+        const data = {
+            plutusTransactionReferenceID: localStorage.getItem("plutusTransactionReferenceID"),
+            amount: line.amount * 100, // We need to provide the amount in paisa since Pine Labs processes amounts in paisa.
```

Comment:
if we go with this approach, please specify in the comment the fact that we assume that the currency is rupees and the relation between rupee and paisa is 1 to 100.

Is there a mechanism that assures us that the pos that uses pinelabs as a payment method has `currency_id.name == "INR"` ?